connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java

```diff
@@ -294,7 +294,12 @@ public synchronized void restartConnector(String connName, Callback<Void> cb) {
 
         worker.stopAndAwaitConnector(connName);
 
-        startConnector(connName, (error, result) -> cb.onCompletion(error, null));
+        startConnector(connName, (error, targetState) -> {
+            if (targetState == TargetState.STARTED) {
+                requestTaskReconfiguration(connName);
```

Comment:
I'm wondering if this should be submitted as a runnable to `requestExecutorService` instead - since `restartConnector` is also called by the `POST /{connector}/restart` API, and this change could lead to a significantly higher response latency. Also, if we were to do that, should we add a `sleep` call to the relevant unit tests and retain the same expectations / verifications (with the assumption that the requestExecutorService will complete the task in the background) ?

Reply:
I think this is fine as-is, we synchronize in `requestTaskReconfiguration` either way. I can't think of a significant difference between invoking that method in the `WorkerConnector`'s thread and invoking it via the herder's `requestExecutorService`.

In fact, I don't think we need to invoke anything via the `requestExecutorService` except in the variant of `restartConnector` that accepts a delay. But that's out of scope for this PR and should be addressed separately in order to simplify review.